inject/hk2/src/main/java/org/glassfish/jersey/inject/hk2/SupplierFactoryBridge.java

```diff
@@ -85,9 +87,11 @@ public T provide() {
     @Override
     public void dispose(T instance) {
         if (disposable) {
-            DisposableSupplier<T> disposableSupplier = disposableSuppliers.get(instance);
-            disposableSupplier.dispose(instance);
-            disposableSuppliers.remove(instance);
+            synchronized (disposableSuppliers) {
+                DisposableSupplier<T> disposableSupplier = disposableSuppliers.get(instance);
```

Comment:
Do not use synchronized block (see previous comment - use synchronized map in field declaration).